### ISSUE TITLE
[ENH]: Will matplotlib support asynchronous functions or generators as `FuncAnimation`'s parameters?

### ISSUE THREAD

**Issue by McDic** [matplotlib/matplotlib#22842]
### Problem

I want to pull data source from web server, database, extra, and nowadays most of them are developed with `asyncio`.

For example, it would be much appreciated if following code works;

```python
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.animation as animation

fig, ax = plt.subplots()

x = np.arange(0, 2 * np.pi, 0.01)
(line,) = ax.plot(x, np.sin(x))


async def animate(i):  # LOOK AT HERE!
    line.set_ydata(np.sin(x + i / 50))  # update the data.
    return (line,)


ani = animation.FuncAnimation(fig, animate, interval=20, blit=True, save_count=50)
plt.show()
```

If `FuncAnimation` accepts asynchronous functions/generators for `func`, `init_func`, `frames`, it would be much better to integrate real time data to plotting.

### Proposed solution

I guess adding something like `event_loop.run_until_completed(coroutine)` is enough but I heard Matplotlib is blocking all event loops for GUI.. So I guess my thought would not be so easy?

**Comment by tacaswell** [comment]
https://matplotlib.org/devdocs/users/explain/interactive_guide.html has a lot of details about event loops and how to integrate them.

For the UI to be responsive (pan / zoom / ect) the GUI main loop needs to be running and processing events.  If you are blocking the main thread waiting on network or database I/O the then everything is locked while we wait.  

If you want to get the concurrency of having your application you either have to integrate your cooperative multi-tasking with the GUI event loop (qumash does this for Qt) or push the Python asyncio event loop to a background thread and use something as a bridge between them.  

My knee-jerk idea of how to deal with the async generator is to have a shared `Queue` with a sync generator that either pops and yields the latest value in the queue or yields the last value it yielded immediately and then a task on a (always running) event loop that pushes to that shared queue.  You give the generator to FuncAnimiation, arrange for the async coro to run in a background task, and then start the GUI event loop.

**Comment by McDic** [comment]
Thank you for your in-depth feedback. Here are some my questions;

> For the UI to be responsive (pan / zoom / ect) the GUI main loop needs to be running and processing events. If you are blocking the main thread waiting on network or database I/O the then everything is locked while we wait.

Maybe because I am not much experienced coder, I don't understand why everything is locked when we do network or database I/O and GUI operation in same loop? Shouldn't they run concurrently when I call like `loop.run_until_complete(coro())` until `coro()` is completed?

> My knee-jerk idea of how to deal with the async generator is to have a shared `Queue` with a sync generator that either pops and yields the latest value in the queue or yields the last value it yielded immediately and then a task on a (always running) event loop that pushes to that shared queue. You give the generator to FuncAnimiation, arrange for the async coro to run in a background task, and then start the GUI event loop.

But isn't it same (everything is blocked) because matplotlib waits to receive the data which is needed to be yielded from synchronous generator anyway?

**Comment by tacaswell** [comment]
We have only just met so I do not yet know what you know, so if I am answering at the wrong level (either talking down to you or over you head) please say so and I'll try to adjust!

> I don't understand why everything is locked when we do network or database I/O and GUI operation in same loop? 

If you do not do anything in particular to get concurrency (threading, multiprocessing, asyncio tasks, ....) code runs serially (which is what makes it possible to reason about it) so `loop.run_until_complete` runs the passed coroutine until it completes and then goes to the next (synchronous) call.  If within `coro` that spawns a new task(s) they will run in parallel with each other, but once `run_until_complete` starts, nothing else runs and once it returns any unfinished tasks that were started are left in suspended limbo (because when you are running the asycio event it is what is managing the cooperative concurency between the tasks (which again only one of them is _actualy_ running at a time but with anything involving I/O it is mostly waiting anyway so you can switch between them)). 

 https://vorpus.org/blog/notes-on-structured-concurrency-or-go-statement-considered-harmful/ has a lot more details and is worth reading.

> But isn't it same (everything is blocked) because matplotlib waits to receive the data which is needed to be yielded from synchronous generator anyway?

Yes and no.  Yes it is true that in syncronous land Matplotlib (and Python and the GUI loop) are blocked waiting for the sync generator to return.  No, because you can write your generator so it always yields quickly.  Consider the following

```python
from queue import Queue, Empty

q = Queue()

def quick_gen(q, initial_val=None):
    last_value = initial_val

    while True:
        try: 
            last_value = q.get(block=False)
        except Empty:
            ...
        yield last_value


g = quick_gen(q, initial_val='bob')

print(next(g))
print(next(g))
q.put('bill')
print(next(g))
print(next(g))
for letter in 'abcde':
    q.put(letter)
for i in range(12):
    print(f'{i}: { next(g)}')
```

In this case I put the `queue.put` calls in-line because starting threads and timers gets messy fast, but that method is thread safe so you can imagine if it were.  Thus, you can have the generator passed into `FuncAnimation` which will pull the next batch of data on a clock (the clock which is implemented via timers from the underlying GUI toolkit is one of the big reasons you want to use the `animatinon` module) and when it asks it will _always_ get an answer promptly. 


Sorry this is a bit rambling...

**Comment by McDic** [comment]
I think it is beyond my level, or maybe my translation is bit incorrect. Please understand if my further question below is stupid..

> but once run_until_complete starts, nothing else runs

I thought unfinished tasks which was spawned before `run_until_complete` is called are still under event loop's management? Like, once the loop is restarted again, then all tasks from suspended limbo restarts again. For example, imagine following code;

```python
import asyncio
import time

async def forever():
    while True:
        await asyncio.sleep(0.5)
        print("forever")

async def spawn_before():
    print("Spawn before")
    task = asyncio.create_task(forever())
    await asyncio.sleep(2)

async def spawn_after():
     print("Spawn after")
     await asyncio.sleep(2)

loop = asyncio.get_event_loop()
loop.run_until_complete(spawn_before())
print("Sleeping 2 seconds..")
time.sleep(2)
loop.run_until_complete(spawn_after())
```

The code above will produce following results; (Event loop resurrects tasks which is stopped)

```
Spawn before
forever
forever
forever
Sleeping 2 seconds..
Spawn after
forever
forever
forever
forever
```

And I guessed many of GUI frameworks' background task involves such event loop, so whenever GUI calls event loop to run their own task, there is a possibility to run other tasks(network or database I/O etc) concurrently. That's why I thought creation of such tasks for other purpose is safe, but from your answer, it seems not..

For your second answer, it is pretty clear and clever!

**Comment by tacaswell** [comment]
> Like, once the loop is restarted again, then all tasks from suspended limbo restarts again

True, but it is best if you can leave the event loop running (or make sure it is restarted frequently).  If you are holding onto one side of a network connection or database cursor and then stay suspended for too long the server side will timeout.  Having the server side cut you off can lead to really odd errors (speaking from experience).

> And I guessed many of GUI frameworks' background task involves such event loop, 

GUI event loops almost always want to run in the foreground on the main thread, but do indeed involve an event loop!  In the asyncio event loop there is a list of tasks who can say "Pick me!  I'm ready to run!" and then they run until they say "right, I'm done, it is someone elses turn" (aka cooperative mutitasking). In the case of GUI event loops they tend to be a queue of work that have events added to it and each piece of work runs until it is complete (typically via a dispatch mechanism so you register callbacks) and then the next event is processed and so on.  Because the UI events from the OS (e.g. mouse / keyboard input!) gets put on the same queue so if you have a long running callback you can effectively lock up the whole UI!  This very quickly leads to the UI frameworks having nice tools for "please put this work on a thread and tell me when it is done", but you have to explicitly opt into these tools (because multithreading is hard!).


> That's why I thought creation of such non-GUI tasks in GUI's event loop is safe, but from your answer, it seems not..

In the case of the animation frame work we trigger the updates via the UI frameworks timer mechanism (you say "dear event loop please run this function about X seconds from now") which is ultimately processed on the main thread as part of the UI event loop.  From the point of view of the event loop the sequence is:

 - Notice the timer is up
 - call your function
 - wait for it to return
 - sort out what to do next

In the "call your function" step the UI has no idea (for very good reason!) what you are doing inside.  You could be doing `time.sleep` or `ev.run_until_complete(...)` or training a ML model, all it knows is if your callback has returned or not.  What you really want to do in the callback is kickoff some process to get more data, tell the UI event loop "when this finishes do this other thing", and then return as quickly as you can.

If you are using Qt you maybe able to use https://pypi.org/project/Quamash/ which runs the asyncio event loop _on_ the Qt event loop so it mostly cooperates as you expect (but you might have to reentrantly kick the event loop? I have to actually used quamash so I may not have the right mental model for what it is doing).

Another option is to manually integrate the two event loops and call one from the other.  One example of this is the "input hook" for the command prompt that runs the GUI event loop while waiting for the human to type which gets you both a responsive UI and a responsive terminal.  Another example is in my BNL work we had a thing called `qt_kicker` that would from within the asyncio event loop would periodically run the Qt event loop to let the UI seem responsive while the asycio event loop was running the show.

> I think it is beyond my level,

I assure you it is beyond everyone's level when they first come to it!  At the end of the days computers do exactly what we tell them to,  but it is not always obvious how all of the things we told them to do are going to interact ;).  There is no magic, only engineering (to be understood).

As you can tell from my somewhat rambling answers despite having spent a fair amount of time working on this sort of thing, I still struggle to give clear and concise explanations.

**Comment by McDic** [comment]
I am pretty much appreciated for your long and detailed explanation! I think your explanation gives lots of information for people who are new to these stuffs like me, so don't think your answer is rambling!

I will code some experimental stuffs to find how to cleverly deliver concurrency (using `quamash`, multithreaded event loops, etc), so I can elegantly integrate my own library([link](https://github.com/McDic/IndicatorManagement/), in case if anyone is interested) and matplotlib.

I am closing this issue. Thanks again for your help!

**Comment by tacaswell** [comment]
Please add the "Framework::Matplotlib" classifier (see https://matplotlib.org/mpl-third-party/ for instructions) to your package on pypi and consider adding it to our list when you feel it is ready for users!
